client/lib/browser-storage/index.ts

```diff
@@ -146,12 +180,39 @@ export async function getStoredItem< T >( key: string ): Promise< T | undefined
 	return await idbGet( key );
 }
 
+/**
+ * Get all stored items.
+ *
+ * @returns A promise with the stored key/value pairs as an object. Empty if none.
+ */
+export async function getAllStoredItems< T >(): Promise< { [ key: string ]: T } > {
```

Comment:
This function (and also `idbGetAll`) should be able to load only keys that match a certain pattern.

In our case, if the user ID is 1234, we want to load keys that are in format
```
redux-state-1234(:[a-z]+)?
```

If the Indexed DB also contains state for user 5678, we should ignore the `redux-state-5678` keys.

There's one special case: if the Signup section cannot find its `redux-state-1234:signup` key, it should load the `redux-state-logged-out:signup` key as a fallback. That's used when signing up a new user and transferring the signup state across the logged-out/logged-in boundary. At some point during the signup, the new user becomes logged-in and the signup continues. And wants to use state stored during the logged-out part.

Reply:
I'm not sure if the `{ [ key: string ]: T }` type is correct. The stored values don't necessarily all have the same type `T`. They are generic objects (results of `JSON.parse`), but can theoretically also be strings or numbers.

Reply:
> This function (and also `idbGetAll`) should be able to load only keys that match a certain pattern.
> 
> In our case, if the user ID is 1234, we want to load keys that are in format
> 
> ```
> redux-state-1234(:[a-z]+)?
> ```
> 
> If the Indexed DB also contains state for user 5678, we should ignore the `redux-state-5678` keys.
> 
> There's one special case: if the Signup section cannot find its `redux-state-1234:signup` key, it should load the `redux-state-logged-out:signup` key as a fallback. That's used when signing up a new user and transferring the signup state across the logged-out/logged-in boundary. At some point during the signup, the new user becomes logged-in and the signup continues. And wants to use state stored during the logged-out part.

Thanks for the detailed explanation, @jsnajdr ! Considering what you just mentioned, it sounds like it's best to keep things simple at the `lib/browser-storage` and accept a regular expression as a pattern to match. The logic can then stay in whatever module is using `lib/browser-storage`.

Reply:
> I'm not sure if the `{ [ key: string ]: T }` type is correct. The stored values don't necessarily all have the same type `T`. They are generic objects (results of `JSON.parse`), but can theoretically also be strings or numbers.

You're right, I don't think that applies here. Created a new type, which unfortunately must contain an explicit `any` 😞

Reply:
Maybe we should be creating a separate `objectStore` for each user id?

Reply:
Not sure how we'd work that with the fallback-to-localStorage part though. 

To elaborate on that, I think partitioning each user into their own objectStore would allow us to cleanly clear each store without having to worry about stepping on others toes. It may also allow us to implement the bypass store in a way that more closely mimics how the normal backing store works and make that path more reliable when testing...

Reply:
I'm not opposed to using a different `objectStore` per user if it helps work around browser issue. 

It wouldn't fit well into the existing architecture, but we can always rework it, either by attempting to merge the two layers, or by adding extra functionality to this layer. As for the `localStorage` fallback, it would handle that as well, with a noop if need be.

Reply:
> You're right, I don't think that applies here. Created a new type, which unfortunately must contain an explicit any 😞

We could create a type which carries through but it'd be useless since these values only inhabit a type at runtime.

If we want to keep the safety up then we might prefer `unknown` to `any` since that will require people to verify that the data they're using matches their expectations.

![Screen Shot 2020-01-16 at 10 12 31 AM](https://user-images.githubusercontent.com/5431237/72546843-babd7280-3848-11ea-934f-ca2f7e2ca7b6.png)
[Playground Link](http://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=2&pc=22#)

Can also give control to the callee to work around the type checks if we say we know exactly what type we expect here.

```ts
async function getMyStoredItems<T extends JSONObject>(pattern?: RegExp): Promise<T> {
 	const results: StoredItemz = {};
 	for ( const [ key, value ] of memoryStore.entries() ) {
 		if ( ! pattern || pattern?.test( key ) ) {
 			results[ key ] = value;
 		}
 	}
 	return results as T;
}

async function mine() {
  const values = await getMyStoredItems<{ name: string }>(/test/);

  // assumes that name exists and is a string
  return values.name.length;
}
```

Reply:
Thanks, @dmsnell! `unknown` looks like a very good option. I wasn't aware of how it was different from `any`, but I've since done some research and it looks like a good fit.